File: terminal-capture-shell-golang-pty-standalone/cmd/new.go
package cmd

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"os/signal"
	"strings"
	"sync"
	"sync/atomic"
	"syscall"
	"time"

	"flashback-shell-pty/pkg/capture"
	"flashback-shell-pty/pkg/config"
	"flashback-shell-pty/pkg/log"
	"flashback-shell-pty/pkg/ptywrap"
	"flashback-shell-pty/pkg/rawmode"
	"flashback-shell-pty/pkg/server"
	"flashback-shell-pty/pkg/vtcapture"
)

// NewCmd handles the "new" subcommand.
func NewCmd(cfg *config.Config, logger *log.Logger, args []string) int {
	var (
		captureFile     string
		captureTextFile string
		vtLogFile       string
		vtLogInterval   = 2 * time.Second
	)

	// Parse known subcommand flags from the front of args, leaving the rest as
	// arguments for the shell. Stop at the first non-flag argument or an
	// explicit "--" separator.
	i := 0
parseLoop:
	for i < len(args) {
		arg := args[i]
		if arg == "--" {
			i++
			break
		}
		if !strings.HasPrefix(arg, "-") {
			break
		}
		// Long flags only.
		if !strings.HasPrefix(arg, "--") {
			// First short-looking argument is treated as a shell argument.
			break
		}
		switch {
		case arg == "--capture":
			if i+1 >= len(args) {
				logger.Errorf("missing value for --capture")
				return 1
			}
			captureFile = args[i+1]
			i += 2
		case strings.HasPrefix(arg, "--capture="):
			captureFile = strings.TrimPrefix(arg, "--capture=")
			i++
		case arg == "--capture-text":
			if i+1 >= len(args) {
				logger.Errorf("missing value for --capture-text")
				return 1
			}
			captureTextFile = args[i+1]
			i += 2
		case strings.HasPrefix(arg, "--capture-text="):
			captureTextFile = strings.TrimPrefix(arg, "--capture-text=")
			i++
		case arg == "--vt-log":
			if i+1 >= len(args) {
				logger.Errorf("missing value for --vt-log")
				return 1
			}
			vtLogFile = args[i+1]
			i += 2
		case strings.HasPrefix(arg, "--vt-log="):
			vtLogFile = strings.TrimPrefix(arg, "--vt-log=")
			i++
		case arg == "--vt-log-interval":
			if i+1 >= len(args) {
				logger.Errorf("missing value for --vt-log-interval")
				return 1
			}
			d, err := time.ParseDuration(args[i+1])
			if err != nil {
				logger.Errorf("invalid --vt-log-interval: %v", err)
				return 1
			}
			vtLogInterval = d
			i += 2
		case strings.HasPrefix(arg, "--vt-log-interval="):
			d, err := time.ParseDuration(strings.TrimPrefix(arg, "--vt-log-interval="))
			if err != nil {
				logger.Errorf("invalid --vt-log-interval: %v", err)
				return 1
			}
			vtLogInterval = d
			i++
		default:
			// Unknown long flag: treat it and the rest as shell args.
			break parseLoop
		}
	}
	shellArgs := args[i:]

	if len(shellArgs) > 0 && strings.TrimLeft(shellArgs[0], "-") == "no-capture" {
		logger.Errorf("--no-capture is a global flag and must appear before the subcommand (e.g. flashback-shell-pty --no-capture new)")
		return 1
	}

	if os.Getenv("FLASHBACK_SHELL") != "" && !cfg.AllowNested {
		logger.Errorf("refusing to start a nested flashback-shell session (FLASHBACK_SHELL is set); use --allow-nested or set allow_nested: true to override")
		return 1
	}

	logger.Infof("effective config: server_url=%s shell=%s buffer_size=%d buffer_mode=%s buffer_dir=%s device_id=%s capture_interval=%d first_capture_delay=%d disable_capture=%t diff_only=%t diff_mode=%s text_only=%t scrollback_lines=%d allow_nested=%t",
		cfg.ServerURL, cfg.Shell, cfg.BufferSize, cfg.BufferMode, cfg.BufferDir, cfg.DeviceID, cfg.CaptureInterval, cfg.FirstCaptureDelay, cfg.DisableCapture, cfg.DiffOnly, cfg.DiffMode, cfg.TextOnly, cfg.ScrollbackLines, cfg.AllowNested)

	shellBin := cfg.Shell
	if shellBin == "" {
		shellBin = os.Getenv("SHELL")
		if shellBin == "" {
			shellBin = "/bin/bash"
		}
	}

	if len(shellArgs) == 0 {
		shellArgs = []string{"-l"}
	}

	cmd := exec.Command(shellBin, shellArgs...)
	cmd.Env = os.Environ()
	if os.Getenv("TERM") == "" {
		cmd.Env = append(cmd.Env, "TERM=xterm-256color")
	}
	cmd.Env = setEnv(cmd.Env, "FLASHBACK_SHELL", "1")

	ptmx, err := ptywrap.Start(cmd)
	if err != nil {
		logger.Errorf("failed to start pty: %v", err)
		return 1
	}
	defer ptmx.Close()

	vtTerm := vtcapture.NewTerminal(80, 24)
	vtTerm.SetScrollbackSize(cfg.ScrollbackLines)

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	var wg sync.WaitGroup

	// Set initial size from the local terminal.
	if rawmode.IsTerminal(int(os.Stdin.Fd())) {
		cols, rows, err := rawmode.Size(int(os.Stdin.Fd()))
		if err == nil && cols > 0 && rows > 0 {
			_ = ptywrap.Resize(ptmx, uint16(rows), uint16(cols))
			vtTerm.Resize(cols, rows)
		}
	}

	// Continuous VT capture log.
	var vtLogger *logBuf
	if vtLogFile != "" {
		f, err := os.Create(vtLogFile)
		if err != nil {
			logger.Errorf("vt log file: %v", err)
			return 1
		}
		defer f.Close()
		vtLogger = newLogBuf(f)
		wg.Add(1)
		go func() {
			defer wg.Done()
			ticker := time.NewTicker(vtLogInterval)
			defer ticker.Stop()
			for {
				select {
				case <-ctx.Done():
					return
				case <-ticker.C:
					capture := vtTerm.CaptureANSI()
					cols, rows := vtTerm.Size()
					vtLogger.Writef("--- %s %dx%d ---\n", time.Now().Format(time.RFC3339Nano), cols, rows)
					vtLogger.Write(capture)
					vtLogger.WriteByte('\n')
					vtLogger.Flush()
				}
			}
		}()
	}

	// Signal handling for clean shutdown. SIGHUP is included so closing the
	// terminal window terminates the wrapper cleanly.
	sigCh := make(chan os.Signal, 1)
	signal.Notify(sigCh, os.Interrupt, syscall.SIGTERM, syscall.SIGHUP)
	defer signal.Stop(sigCh)

	// Raw mode.
	var restore func() error
	if rawmode.IsTerminal(int(os.Stdin.Fd())) {
		var err error
		restore, _, err = rawmode.MakeRaw(int(os.Stdin.Fd()))
		if err != nil {
			logger.Warnf("make raw: %v", err)
		}
	}
	if restore != nil {
		defer restore()
	}

	// Resize monitor.
	resizeCh := make(chan os.Signal, 1)
	signal.Notify(resizeCh, syscall.SIGWINCH)
	defer signal.Stop(resizeCh)
	wg.Add(1)
	go func() {
		defer wg.Done()
		for {
			select {
			case <-ctx.Done():
				return
			case <-resizeCh:
				if rawmode.IsTerminal(int(os.Stdin.Fd())) {
					cols, rows, err := rawmode.Size(int(os.Stdin.Fd()))
					if err == nil && cols > 0 && rows > 0 {
						_ = ptywrap.Resize(ptmx, uint16(rows), uint16(cols))
						vtTerm.Resize(cols, rows)
						logger.Debugf("resized to %dx%d", cols, rows)
					}
				}
			}
		}
	}()

	// Forward stdin -> PTY. This goroutine is intentionally not part of the
	// wait group: os.Stdin.Read can stay blocked after the child exits, and
	// waiting for it would prevent clean shutdown.
	go func() {
		buf := make([]byte, 4096)
		for {
			n, err := os.Stdin.Read(buf)
			if n > 0 {
				_, _ = ptmx.Write(buf[:n])
			}
			if err != nil {
				return
			}
		}
	}()

	// Counters for VT emulator health.
	var vtDroppedBytes atomic.Int64
	var vtForwardedBytes atomic.Int64

	// Forward PTY -> stdout and VT emulator.
	// VT processing is async with a timeout so that a blocked emulator
	// (e.g. its reply pipe full because nobody consumes it) does not freeze
	// live stdout output.
	wg.Add(1)
	go func() {
		defer wg.Done()
		vtInput := make(chan []byte, 256)

		wg.Add(1)
		go func() {
			defer wg.Done()
			for data := range vtInput {
				done := make(chan struct{})
				go func() {
					_, _ = vtTerm.Write(data)
					close(done)
				}()
				select {
				case <-done:
				case <-time.After(500 * time.Millisecond):
					logger.Warnf("VT emulator write timed out, disabling VT captures")
					for range vtInput {
					}
					return
				}
			}
		}()

		buf := make([]byte, 4096)
		for {
			n, err := ptmx.Read(buf)
			if n > 0 {
				_, _ = os.Stdout.Write(buf[:n])
				data := append([]byte(nil), buf[:n]...)
				select {
				case vtInput <- data:
				default:
					vtDroppedBytes.Add(int64(n))
				}
			}
			if err != nil {
				close(vtInput)
				return
			}
		}
	}()

	// VT emulator input pipe -> PTY.
	// The emulator writes reply sequences (e.g. OSC 10/11 color responses) to
	// an internal pipe; draining it and forwarding those bytes back to the PTY
	// prevents TUIs like vim from blocking on terminal queries.
	wg.Add(1)
	go func() {
		defer wg.Done()
		buf := make([]byte, 4096)
		for {
			n, err := vtTerm.Read(buf)
			if n > 0 {
				vtForwardedBytes.Add(int64(n))
				_, _ = ptmx.Write(buf[:n])
			}
			if err != nil {
				return
			}
		}
	}()

	// Background capture and upload.
	var captureCancel context.CancelFunc
	var captureWg sync.WaitGroup
	if !cfg.DisableCapture && cfg.CaptureInterval > 0 {
		var captureCtx context.Context
		captureCtx, captureCancel = context.WithCancel(ctx)
		captureWg.Add(1)
		go func() {
			defer captureWg.Done()
			backgroundCapture(captureCtx, cfg, vtTerm, logger)
		}()
	}

	// Wait for command exit in a goroutine so we can react to signals.
	waitDone := make(chan struct{})
	var exitErr error
	go func() {
		exitErr = cmd.Wait()
		close(waitDone)
	}()

	select {
	case sig := <-sigCh:
		logger.Infof("received signal %s, shutting down", sig)
		_ = cmd.Process.Kill()
		<-waitDone
	case <-waitDone:
	}

	status := 0
	if exitErr != nil {
		if ee, ok := exitErr.(*exec.ExitError); ok {
			status = ee.ExitCode()
		} else {
			status = 1
		}
	}

	if captureCancel != nil {
		captureCancel()
		captureWg.Wait()
	}

	// Final capture writes.
	if captureFile != "" {
		if err := os.WriteFile(captureFile, vtTerm.CaptureANSI(), 0644); err != nil {
			logger.Warnf("failed to write ansi capture: %v", err)
		}
	}
	if captureTextFile != "" {
		if err := os.WriteFile(captureTextFile, vtTerm.CaptureText(), 0644); err != nil {
			logger.Warnf("failed to write text capture: %v", err)
		}
	}

	if vtLogger != nil {
		vtLogger.Flush()
	}

	cancel()
	// Closing the VT emulator unblocks the input-pipe forwarder goroutine so
	// wg.Wait() can return promptly.
	_ = vtTerm.Close()
	wg.Wait()

	return status
}

func backgroundCapture(ctx context.Context, cfg *config.Config, vtTerm *vtcapture.Terminal, logger *log.Logger) {
	buffer, cleanup, err := newRetryBuffer(cfg)
	if err != nil {
		logger.Warnf("failed to create retry buffer: %v", err)
	}
	if cleanup != nil {
		defer cleanup()
	}

	client := server.NewClient(cfg, buffer, logger)

	// Flush any retries from a previous run. With the temp subfolder pattern,
	// previous data is already gone, so this is effectively a no-op.
	if buffer != nil {
		if err := client.FlushRetries(); err != nil {
			logger.Debugf("flush retries: %v", err)
		}
	}

	engine := capture.NewEngine("")
	engine.DiffOnly = cfg.DiffOnly
	engine.DiffMode = cfg.DiffMode
	engine.TextOnly = cfg.TextOnly

	do := func() {
		c := engine.CaptureScreen(vtTerm, cfg)
		if c == nil {
			return
		}
		logger.Infof("background capture: changed")
		if err := client.Upload([]capture.Capture{*c}); err != nil {
			logger.Debugf("background upload: %v", err)
		}
	}

	interval := time.Duration(cfg.CaptureInterval) * time.Second
	logger.Infof("background capture started (first_capture_delay=%s, interval=%s)", time.Duration(cfg.FirstCaptureDelay)*time.Second, interval)

	if cfg.FirstCaptureDelay > 0 {
		select {
		case <-ctx.Done():
			logger.Infof("background capture stopped")
			return
		case <-time.After(time.Duration(cfg.FirstCaptureDelay) * time.Second):
		}
	}

	do()

	ticker := time.NewTicker(interval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			logger.Infof("background capture stopped")
			return
		case <-ticker.C:
			do()
		}
	}
}

func newRetryBuffer(cfg *config.Config) (*capture.Buffer, func(), error) {
	parent := cfg.BufferDir
	if parent == "" {
		parent = "/tmp"
	}
	dir, err := os.MkdirTemp(parent, ".flashback-shell-pty-buffer-*")
	if err != nil {
		return nil, nil, err
	}
	cleanup := func() {
		_ = os.RemoveAll(dir)
	}
	return capture.NewBuffer(dir, cfg.BufferSize), cleanup, nil
}

// logBuf wraps an os.File with a small buffer.
type logBuf struct {
	w   *os.File
	buf []byte
}

func newLogBuf(f *os.File) *logBuf {
	return &logBuf{w: f}
}

func (l *logBuf) Writef(format string, a ...interface{}) {
	l.buf = append(l.buf, []byte(fmt.Sprintf(format, a...))...)
}

func (l *logBuf) Write(p []byte) {
	l.buf = append(l.buf, p...)
}

func (l *logBuf) WriteByte(b byte) error {
	l.buf = append(l.buf, b)
	return nil
}

func (l *logBuf) Flush() {
	if len(l.buf) == 0 {
		return
	}
	_, _ = l.w.Write(l.buf)
	l.buf = l.buf[:0]
}

func setEnv(env []string, key, value string) []string {
	prefix := key + "="
	for i, e := range env {
		if strings.HasPrefix(e, prefix) {
			env[i] = prefix + value
			return env
		}
	}
	return append(env, prefix+value)
}

File: terminal-capture-shell-rust-standalone/src/cmd/new.rs
use std::collections::HashMap;
use std::path::PathBuf;
use std::process::ExitCode;
use std::sync::Arc;
use std::time::Duration;

use nix::pty::Winsize;
use nix::sys::wait::WaitStatus;
use tokio::signal::unix::{signal, SignalKind};
use tokio::sync::Mutex;
use tokio::time::{interval, sleep};

use crate::capture::buffer::{Buffer, MemoryBuffer, RetryBuffer};
use crate::capture::engine::Engine;
use crate::cli::NewArgs;
use crate::config::{Config, ConfigSources};
use crate::log::Logger;
use crate::pty;
use crate::server::Client;
use crate::tty::DevTty;
use crate::vt_capture::VtCapture;

pub async fn run(
    config: Config,
    _sources: ConfigSources,
    args: NewArgs,
    logger: Logger,
) -> ExitCode {
    if !config.allow_nested && std::env::var("FLASHBACK_SHELL").is_ok() {
        logger.error(
            "refusing to start a nested capture session (FLASHBACK_SHELL is set); use --allow-nested or set allow_nested: true to override",
        );
        return ExitCode::from(1);
    }

    logger.info(&format!(
        "effective config: server_url={} shell={} buffer_size={} buffer_mode={} buffer_dir={} device_id={} capture_interval={} first_capture_delay={} disable_capture={} diff_only={} diff_mode={:?} text_only={} scrollback_lines={} allow_nested={}",
        config.server_url,
        config.shell_binary(),
        config.buffer_size,
        config.buffer_mode,
        config.buffer_dir.display(),
        config.device_id,
        config.capture_interval,
        config.first_capture_delay,
        config.disable_capture,
        config.diff_only,
        config.diff_mode,
        config.text_only,
        config.scrollback_lines,
        config.allow_nested,
    ));

    let shell_bin = config.shell_binary();
    let shell_args = if args.shell_args.is_empty() {
        vec!["-l".to_owned()]
    } else {
        args.shell_args
    };
    let command = vec![shell_bin]
        .into_iter()
        .chain(shell_args.into_iter())
        .collect::<Vec<_>>();

    // Open /dev/tty for raw I/O and size.
    let tty_reader = match DevTty::open().await {
        Ok(t) => t,
        Err(e) => {
            logger.error(&format!("failed to open /dev/tty: {e}"));
            return ExitCode::from(1);
        }
    };
    let tty_writer = match tty_reader.try_clone() {
        Ok(t) => t,
        Err(e) => {
            logger.error(&format!("failed to clone /dev/tty: {e}"));
            return ExitCode::from(1);
        }
    };
    let tty_resizer = match tty_reader.try_clone() {
        Ok(t) => t,
        Err(e) => {
            logger.error(&format!("failed to clone /dev/tty: {e}"));
            return ExitCode::from(1);
        }
    };
    let size = tty_reader.get_size();
    let (cols, rows) = if size.ws_col > 0 && size.ws_row > 0 {
        (size.ws_col, size.ws_row)
    } else {
        logger.warn("tty reported 0x0, using default size 80x24");
        (80, 24)
    };

    // Spawn PTY.
    let winsize = Winsize {
        ws_col: cols,
        ws_row: rows,
        ws_xpixel: 0,
        ws_ypixel: 0,
    };
    let mut extra_env = HashMap::<String, String>::new();
    extra_env.insert("FLASHBACK_SHELL".to_owned(), "1".to_owned());
    if std::env::var("TERM").is_err() {
        extra_env.insert("TERM".to_owned(), "xterm-256color".to_owned());
    }

    let pty = match pty::spawn(&command, winsize, &extra_env) {
        Ok(p) => Arc::new(p),
        Err(e) => {
            logger.error(&format!("failed to start pty: {e}"));
            return ExitCode::from(1);
        }
    };

    // VT emulator.
    let vt = Arc::new(Mutex::new(VtCapture::new(
        cols as usize,
        rows as usize,
        config.scrollback_lines,
    )));

    // Broadcast channel for background-task shutdown.
    let (shutdown_tx, _) = tokio::sync::broadcast::channel::<()>(1);

    // VT log task.
    let vt_log_task = if let Some(vt_log_path) = args.vt_log_file {
        let vt = vt.clone();
        let interval_duration = args.vt_log_interval;
        let logger = logger.clone();
        let mut shutdown = shutdown_tx.subscribe();
        Some(tokio::spawn(async move {
            let mut ticker = interval(interval_duration);
            loop {
                tokio::select! {
                    _ = shutdown.recv() => break,
                    _ = ticker.tick() => {}
                }
                let (ansi, cols, rows) = {
                    let vt = vt.lock().await;
                    let (cols, rows) = vt.size();
                    (vt.capture_ansi(), cols as u16, rows as u16)
                };
                let line = format!(
                    "--- {} {}x{} ---\n{}\n",
                    chrono::Local::now().to_rfc3339(),
                    cols,
                    rows,
                    ansi
                );
                if let Err(e) = tokio::fs::write(&vt_log_path, line).await {
                    logger.warn(&format!("failed to write vt log: {e}"));
                }
            }
        }))
    } else {
        None
    };

    // Background capture.
    let (retry_buffer, retry_cleanup) = new_retry_buffer(&config);
    let capture_task = if !config.disable_capture && config.capture_interval > 0 {
        let config = config.clone();
        let vt = vt.clone();
        let logger = logger.clone();
        let shutdown = shutdown_tx.subscribe();
        let client = Client::new(
            config.server_url.clone(),
            config.device_id.clone(),
            retry_buffer,
            logger.clone(),
        );
        Some(tokio::spawn(async move {
            background_capture(config, vt, logger, shutdown, client).await;
        }))
    } else {
        None
    };

    // Wait for child exit in a blocking task.
    let pty_for_wait = pty.clone();
    let (exit_tx, mut exit_rx) = tokio::sync::watch::channel::<Option<WaitStatus>>(None);
    tokio::spawn(async move {
        let status = match pty_for_wait.wait(None).await {
            Ok(status) => Some(status),
            Err(_) => None,
        };
        let _ = exit_tx.send(status);
    });

    // Signals.
    let mut sigint = match signal(SignalKind::interrupt()) {
        Ok(s) => s,
        Err(e) => {
            logger.error(&format!("failed to register SIGINT handler: {e}"));
            return ExitCode::from(1);
        }
    };
    let mut sigterm = match signal(SignalKind::terminate()) {
        Ok(s) => s,
        Err(e) => {
            logger.error(&format!("failed to register SIGTERM handler: {e}"));
            return ExitCode::from(1);
        }
    };
    let mut sighup = match signal(SignalKind::hangup()) {
        Ok(s) => s,
        Err(e) => {
            logger.error(&format!("failed to register SIGHUP handler: {e}"));
            return ExitCode::from(1);
        }
    };
    let mut sigwinch = match signal(SignalKind::window_change()) {
        Ok(s) => s,
        Err(e) => {
            logger.error(&format!("failed to register SIGWINCH handler: {e}"));
            return ExitCode::from(1);
        }
    };

    // Input bridge: /dev/tty -> PTY.
    let logger_input = logger.clone();
    let pty_for_input = pty.clone();
    let input_bridge = tokio::spawn(async move {
        let mut buf = [0u8; 4096];
        let mut total = 0usize;
        loop {
            match tty_reader.read(&mut buf).await {
                Ok(0) => break,
                Ok(n) => {
                    total += n;
                    if let Err(e) = pty_for_input.write(&buf[..n]).await {
                        logger_input.debug(&format!("pty write error: {e}"));
                        break;
                    }
                }
                Err(e) => {
                    logger_input.debug(&format!("tty read error: {e}"));
                    break;
                }
            }
        }
        logger_input.debug(&format!("input bridge finished after {total} bytes"));
    });

    // Output bridge: PTY -> /dev/tty + VT.
    let logger_output = logger.clone();
    let pty_for_output = pty.clone();
    let vt_for_output = vt.clone();
    let output_bridge = tokio::spawn(async move {
        let mut buf = [0u8; 4096];
        let mut total = 0usize;
        loop {
            match pty_for_output.read(&mut buf).await {
                Ok(0) => {
                    logger_output.debug("pty read returned 0; shell exited");
                    break;
                }
                Ok(n) => {
                    total += n;
                    if let Err(e) = tty_writer.write_all(&buf[..n]).await {
                        logger_output.debug(&format!("tty write error: {e}"));
                        break;
                    }
                    let text = String::from_utf8_lossy(&buf[..n]);
                    vt_for_output.lock().await.feed(&text);
                }
                Err(e) => {
                    logger_output.debug(&format!("pty read error: {e}"));
                    break;
                }
            }
        }
        logger_output.debug(&format!("output bridge finished after {total} bytes"));
    });

    // Resize handler.
    let pty_for_resize = pty.clone();
    let vt_for_resize = vt.clone();
    let logger_resize = logger.clone();
    let resize_task = tokio::spawn(async move {
        loop {
            sigwinch.recv().await;
            let size = tty_resizer.get_size();
            if size.ws_col > 0 && size.ws_row > 0 {
                let winsize = Winsize {
                    ws_col: size.ws_col,
                    ws_row: size.ws_row,
                    ws_xpixel: 0,
                    ws_ypixel: 0,
                };
                pty_for_resize.resize(winsize);
                vt_for_resize
                    .lock()
                    .await
                    .resize(size.ws_col as usize, size.ws_row as usize);
                logger_resize.debug(&format!(
                    "resized to {}x{}",
                    size.ws_col, size.ws_row
                ));
            }
        }
    });

    // Wait for exit or signal.
    let exit_status = tokio::select! {
        _ = sigint.recv() => {
            logger.info("received SIGINT, shutting down");
            pty.kill();
            None
        }
        _ = sigterm.recv() => {
            logger.info("received SIGTERM, shutting down");
            pty.kill();
            None
        }
        _ = sighup.recv() => {
            logger.info("received SIGHUP, shutting down");
            pty.kill();
            None
        }
        _ = exit_rx.changed() => {
            logger.info("shell process exited");
            *exit_rx.borrow()
        }
    };

    // If a signal triggered shutdown, wait briefly for the child to exit.
    let exit_status = match exit_status {
        Some(s) => Some(s),
        None => match tokio::time::timeout(Duration::from_secs(5), exit_rx.changed()).await {
            Ok(Ok(())) => *exit_rx.borrow(),
            _ => {
                logger.warn("timed out waiting for shell to exit after signal");
                None
            }
        },
    };

    let exit_code = exit_status.map(exit_code_for_status).unwrap_or(ExitCode::SUCCESS);

    // Let output bridge drain.
    let _ = tokio::time::timeout(Duration::from_secs(1), output_bridge).await;
    input_bridge.abort();
    resize_task.abort();

    // Cancel background capture and VT log.
    let _ = shutdown_tx.send(());
    if let Some(t) = capture_task {
        let _ = tokio::time::timeout(Duration::from_secs(2), t).await;
    }
    if let Some(cleanup) = retry_cleanup {
        cleanup();
    }
    if let Some(t) = vt_log_task {
        let _ = tokio::time::timeout(Duration::from_secs(1), t).await;
    }

    // Final capture writes.
    let final_ansi = {
        let vt = vt.lock().await;
        vt.capture_ansi()
    };
    let final_text = {
        let vt = vt.lock().await;
        vt.capture_text().join("\n")
    };

    if let Some(path) = args.capture_file {
        if let Err(e) = tokio::fs::write(&path, &final_ansi).await {
            logger.warn(&format!("failed to write ansi capture: {e}"));
        }
    }
    if let Some(path) = args.capture_text_file {
        if let Err(e) = tokio::fs::write(&path, &final_text).await {
            logger.warn(&format!("failed to write text capture: {e}"));
        }
    }

    exit_code
}

fn exit_code_for_status(status: WaitStatus) -> ExitCode {
    match status {
        WaitStatus::Exited(_, code) => ExitCode::from(code as u8),
        WaitStatus::Signaled(_, sig, _) => ExitCode::from(128 + sig as u8),
        _ => ExitCode::SUCCESS,
    }
}

async fn background_capture(
    config: Config,
    vt: Arc<Mutex<VtCapture>>,
    logger: Logger,
    mut shutdown: tokio::sync::broadcast::Receiver<()>,
    client: Client,
) {
    if let Err(e) = client.flush_retries().await {
        logger.debug(&format!("flush retries: {e}"));
    }

    let mut engine = Engine::new(PathBuf::new());
    engine.diff_only = config.diff_only;
    engine.diff_mode = config.diff_mode;
    engine.text_only = config.text_only;

    logger.info(&format!(
        "background capture started (first_capture_delay={}s, interval={}s)",
        config.first_capture_delay, config.capture_interval
    ));

    if config.first_capture_delay > 0 {
        tokio::select! {
            _ = shutdown.recv() => {
                logger.info("background capture stopped");
                return;
            }
            _ = sleep(Duration::from_secs(config.first_capture_delay)) => {}
        }
    }

    do_capture(
        &mut engine,
        &client,
        &vt,
        &logger,
    )
    .await;

    let mut ticker = interval(Duration::from_secs(config.capture_interval));
    loop {
        tokio::select! {
            _ = shutdown.recv() => {
                logger.info("background capture stopped");
                return;
            }
            _ = ticker.tick() => {
                do_capture(
                    &mut engine,
                    &client,
                    &vt,
                    &logger,
                )
                .await;
            }
        }
    }
}

async fn do_capture(
    engine: &mut Engine,
    client: &Client,
    vt: &Arc<Mutex<VtCapture>>,
    logger: &Logger,
) {
    let (ansi, text, cols, rows) = {
        let vt = vt.lock().await;
        (
            vt.capture_ansi(),
            vt.capture_text().join("\n"),
            vt.size().0,
            vt.size().1,
        )
    };

    if let Some(capture) = engine.process("main", "main", "main", ansi, text, cols, rows) {
        logger.info("background capture: changed");
        match client.upload(std::slice::from_ref(&capture)).await {
            Ok(()) => {
                if let Err(e) = engine.save_hashes(std::slice::from_ref(&capture)) {
                    logger.warn(&format!("failed to save hash: {e}"));
                }
            }
            Err(e) => {
                logger.debug(&format!("background upload: {e}"));
            }
        }
    }
}

fn new_retry_buffer(
    config: &Config,
) -> (Option<RetryBuffer>, Option<Box<dyn FnOnce()>>) {
    if config.buffer_mode.to_lowercase() == "disk" {
        let parent = &config.buffer_dir;
        let dir = match std::fs::create_dir_all(parent)
            .and_then(|_| tempfile::tempdir_in(parent))
        {
            Ok(d) => {
                let path = d.path().to_path_buf();
                std::mem::forget(d);
                path
            }
            Err(e) => {
                eprintln!(
                    "warning: cannot create disk buffer dir in {}: {e}; using memory",
                    parent.display()
                );
                return (
                    Some(RetryBuffer::Memory(MemoryBuffer::new(config.buffer_size))),
                    None,
                );
            }
        };
        let cleanup: Box<dyn FnOnce()> = {
            let dir = dir.clone();
            Box::new(move || {
                let _ = std::fs::remove_dir_all(&dir);
            })
        };
        (
            Some(RetryBuffer::Disk(Buffer::new(dir, config.buffer_size))),
            Some(cleanup),
        )
    } else {
        (
            Some(RetryBuffer::Memory(MemoryBuffer::new(config.buffer_size))),
            None,
        )
    }
}
